astropy/table/serialize.py

```diff
@@ -336,7 +336,12 @@ def _construct_mixin_from_columns(new_name, obj_attrs, out):
     info = obj_attrs.pop('__info__', {})
     if len(data_attrs_map) == 1:
         # col is the first and only serialized column; in that case, use info
-        # stored on the column.
+        # stored on the column. First step is to get that first column which
+        # has been moved from `out` to `obj_attrs` above.
+        data_attr = next(iter(data_attrs_map.values()))
```

Comment:
This worked before because `data_attr` is already defined correctly. Somewhat annoying to re-do it just because of the type checker. Maybe again define a default on top instead?

Reply:
In the code above (line 331 loop) `col` is created as a "side-effect" of the actual goal. I actually changed the code to not make that side-effect variable, so it is clean and understandable now. 
```
    for name, data_attr in data_attrs_map.items():
        obj_attrs[data_attr] = out[name]
        del out[name]
```